ra/ra.go

```diff
@@ -751,28 +749,96 @@ func domainsForRateLimiting(names []string) ([]string, error) {
 	return domains, nil
 }
 
+// suffixesForRateLimiting transforms a list of FQDNs into a list of exact
+// public suffix matches for the purpose of rate limiting. It also
+// de-duplicates the output domains. Domains that are not an exact match for a
+// public suffix are not included.
+func suffixesForRateLimiting(names []string) ([]string, error) {
+	domainsMap := make(map[string]struct{}, len(names))
+	var suffixMatches []string
+	for _, name := range names {
+		_, err := publicsuffix.Domain(name)
+		if err != nil {
+			// Like `domainsForRateLimiting`, the only possible errors here are:
+			// (1) publicsuffix.Domain is giving garbage values
+			// (2) the public suffix is the domain itself
+			// We assume 2 and collect it into the result
+			if _, ok := domainsMap[name]; !ok {
+				domainsMap[name] = struct{}{}
+				suffixMatches = append(suffixMatches, name)
+			}
+		}
+	}
+	return suffixMatches, nil
+}
+
+// certCountRPC abstracts the choice of the SA.CountCertificatesByExactNames or
+// the SA.CountCertificatesByNames RPC.
+type certCountRPC func(ctx context.Context, names []string, earliest, lastest time.Time) ([]*sapb.CountByNames_MapElement, error)
+
+// enforceNameCounts uses the provided count RPC to find a count of certificates
+// for each of the names. If the count for any of the names exceeds the limit
+// for the given registration then the names out of policy are returned to be
+// used for a rate limit error.
+func (ra *RegistrationAuthorityImpl) enforceNameCounts(
+	ctx context.Context,
+	names []string,
+	limit ratelimit.RateLimitPolicy,
+	regID int64,
+	countFunc certCountRPC) ([]string, error) {
+
+	now := ra.clk.Now()
+	windowBegin := limit.WindowBegin(now)
+	counts, err := countFunc(ctx, names, windowBegin, now)
+	if err != nil {
+		return nil, err
+	}
+
+	var badNames []string
+	for _, entry := range counts {
+		if int(*entry.Count) >= limit.GetThreshold(*entry.Name, regID) {
```

Comment:
nit: we check the slice is not `nil` in the client wrapper but not that the contents is not, I'm not sure what we do elsewhere but is it worthwhile doing a `nil` check here before dereferencing?